helix-term/src/handlers/diagnostics.rs

```diff
@@ -21,4 +36,264 @@ pub(super) fn register_hooks(_handlers: &Handlers) {
         }
         Ok(())
     });
+
+    let tx = handlers.pull_diagnostics.clone();
+    let tx_all_documents = handlers.pull_all_documents_diagnostics.clone();
+    register_hook!(move |event: &mut DocumentDidChange<'_>| {
+        if event
+            .doc
+            .has_language_server_with_feature(LanguageServerFeature::PullDiagnostics)
+            && !event.ghost_transaction
+        {
+            // Cancel the ongoing request, if present.
+            event.doc.pull_diagnostic_controller.cancel();
+            let document_id = event.doc.id();
+            send_blocking(&tx, PullDiagnosticsEvent { document_id });
+            send_blocking(&tx_all_documents, PullAllDocumentsDiagnosticsEvent {});
+        }
+        Ok(())
+    });
+
+    register_hook!(move |event: &mut DocumentDidOpen<'_>| {
+        request_document_diagnostics(event.editor, event.doc, false);
+
+        Ok(())
+    });
+
+    register_hook!(move |event: &mut LanguageServerInitialized<'_>| {
+        let doc_ids: Vec<_> = event.editor.documents().map(|doc| doc.id()).collect();
+
+        for doc_id in doc_ids {
+            request_document_diagnostics(event.editor, doc_id, false);
+        }
+
+        Ok(())
+    });
+}
+
+#[derive(Debug)]
+pub(super) struct PullDiagnosticsHandler {
+    document_ids: HashSet<DocumentId>,
+}
+
+impl PullDiagnosticsHandler {
+    pub fn new() -> Self {
+        PullDiagnosticsHandler {
+            document_ids: Default::default(),
```

Comment:
Since this is basically `Default`, let's remove the `::new` constructor here, add `Default` to the `#[derive(..)]` for this type, and construct it with `Default::default()` or `PullDiagnosticHandler::default()` anywhere `::new` is called now